### DISCUSSION TITLE
Representing Graphs with Mixed Edges

### DISCUSSION THREAD

**Discussion by adam2392** [networkx/networkx#5408] [Q&A] [answered]
Hi,

I am new to networkx and interested in using it for representation for graphs with mixed edges (i.e. directed edge, bidirected edge, or other types of edges).

Other types of edges may be possible, each with a different meaning. For example in what is known as a Partial Ancestral Graph (PAG), there can be:

- arrowheads (<, >)
- tails (-)
- circles (o)
- stars (*)

such as: ->, <->, o-o

Is there a way to leverage networkx to represent these kinds of graphs?

**Comment by dschult** [comment, marked as answer]
I think that the edge attributes are the right way to encode that information.  Each edge can store whatever info you'd like about its properties.  I would suggest using a DiGraph where two edges in opposite directions can nicely indicate a bidirectional edge. Then add whatever attributes you need to the edge.  That hard part comes when you try to get an algorithm to measure something about those edge attributes. But maybe that is not so bad either...  depends what you want to do. :}

**Reply by adam2392** [reply]
Hmm I see, thanks for the potential solution @dschult! I'm guessing you mean a MultiDiGraph if I need to have the ability to have: `X -> Y` and `X <-> Y`? I'm trying to weight the pros/cons of inheriting from `nx.DiGraph/nx.MultiDiGraph`, or defining a new class altogether. 

What are then classes of algorithms that would not be valid, when you expand the types of edges you can have via "edge attributes"? 

I'm specifically using it for causal inference research. So things like d-separation, sorting a node list topologically, getting parents and children (note parents/children shouldn't go through a bidirected arrow). I wouldn't say... need "in/out/degree" to work necessarily out of the box since that is not well-defined. This is just to encode what's known as a causal DAG (where bidirectional edges representing latent variables). 

I would imagine each of those operations adds some "computational complexity" since they now need to check attributes in addition? Although... maybe it shouldn't be too bad considering that's all within a lookup table (e.g. 4 hashes through a series of dict of dict of dict of dict)?

**Reply by dschult** [reply]
In order to use e.g. the NetworkX `topological_sort()` you will need to have a directed acyclic graph (DAG), so I imagine you would "simply" remove the latent variables (drop the bidirectional edges) and use the provided function. I don't know of any algorithms that actually use a Mixed-Graph data structure. So maybe you create that structure as a way to quickly construct the graphs and digraphs you need for algorithms. If that is the case, I would suggest starting with MultiDiGraph and add methods that construct DiGraphs based on commonly used filters of edges based on attributes.  Something like:  `G.only_bidirected()` or `G.remove_bidirected()`.

**Reply by adam2392** [reply]
I'm not super familiar with networkx, but am reading the code/docs whenever I can :p.

So your suggestion is some helper method perhaps that allows me to easily query only the normal edges, and then also only query the bidirected edges. Then I could "enable" various algorithms allowed for `DiGraph`/`MultiDiGraph`. Then all other algorithms, simply document that using networkx out of the box would break without first calling something like `G.only_directed`?

**Comment by asha24choudhary** [comment]
@adam2392 Did u already manage something. I am trying to achieve the same thing as you. I have applied the FCI causal discovery algorithm. This you can find [here](https://causal-learn.readthedocs.io/en/latest/search_methods_index/Constraint-based%20causal%20discovery%20methods/FCI.html). I am using the adjacency matrix to create the graph using networkx library. However, I see that I am getting two arrays from A ->B & B ->A. I wanna have ->, <->, o-o & o-> types. Please let me know if I can refer something or if there was something u found out?

**Reply by adam2392** [reply]
We've implemented Mixed-edge-graphs in pywhy-graphs: https://github.com/py-why/pywhy-graphs. We move away from using an array to represent things for 3 reasons: i) it breaks from networkx API, ii) it's not really scalable as you need to store all the 0s as well and iii) because you break networkx API, you can't really use any of the algorithms there.

E.g. https://www.pywhy.org/pywhy-graphs/dev/auto_examples/intro/intro_causal_graphs.html#sphx-glr-auto-examples-intro-intro-causal-graphs-py

We also have the FCI algorithm implemented, which leverages the pywhy-graphs framework: https://github.com/py-why/dodiscover

https://www.pywhy.org/dodiscover/dev/auto_examples/plot_pc_alg.html#sphx-glr-auto-examples-plot-pc-alg-py shows the PC, but FCI follows the same API. Note the dodiscover API is not stable at all right now, but it's in active development.
